File: platform/docs/samples/first-payment-swift/Sources/FirstPayment/FirstPayment.swift
import Foundation
import LayerXSDK
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

let settledStates: Set<String> = ["done", "done-finalised", "refused"]
let completedStates: Set<String> = ["done", "done-finalised"]

func fail(_ reason: String) -> Never {
    FileHandle.standardError.write(Data("first-payment-swift: \(reason)\n".utf8))
    exit(1)
}

func required(_ name: String) -> String {
    guard let value = ProcessInfo.processInfo.environment[name], !value.isEmpty else {
        fail("missing \(name)")
    }
    return value
}

func emit(_ value: [String: JSONValue]) {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys]
    guard let encoded = try? encoder.encode(JSONValue.object(value)) else {
        fail("could not encode the report")
    }
    FileHandle.standardOutput.write(encoded)
    FileHandle.standardOutput.write(Data("\n".utf8))
}

func state(of journey: JSONValue) -> String {
    journey.objectValue?["state"]?.stringValue ?? ""
}

func receipts(in journey: JSONValue) -> [JSONValue] {
    guard case let .array(evidence)? = journey.objectValue?["evidence"] else { return [] }
    return evidence.compactMap { entry in
        guard entry.objectValue?["class"]?.stringValue == "layerx-receipt",
              let identifier = entry.objectValue?["evidence_id"]?.stringValue else { return nil }
        return .string(identifier)
    }
}

@main
struct FirstPayment {
    static func main() async {
        guard let serviceURL = URL(string: required("LAYERX_API_URL")) else {
            fail("invalid LAYERX_API_URL")
        }
        let apiToken = required("LAYERX_API_TOKEN")
        let source = required("LAYERX_SOURCE")
        let destination = required("LAYERX_DESTINATION")
        let money = JSONValue.object([
            "amount": .string(required("LAYERX_AMOUNT")),
            "currency": .string(required("LAYERX_CURRENCY")),
        ])
        guard let paymentKey = IdempotencyKey(rawValue: required("LAYERX_PAYMENT_KEY")) else {
            fail("invalid LAYERX_PAYMENT_KEY")
        }

        do {
            // layerx:begin integration
            let token = try AccessToken(Data(apiToken.utf8))
            let layerx = PlatformClient(transport: try HumanHTTPTransport(baseURL: serviceURL, accessToken: token))
            let quote = try await layerx.humanMoveQuote(.object(["source": .string(source), "destination": .string(destination), "money": money]))
            guard let quoteID = quote.objectValue?["quote_id"]?.stringValue else { fail("move quote omitted quote_id") }
            var journey = try await layerx.humanMoveCommit(.object(["quote_id": .string(quoteID)]), idempotencyKey: paymentKey)
            // layerx:end integration

            guard let journeyID = journey.objectValue?["journey_id"]?.stringValue else {
                fail("commit omitted journey_id")
            }
            var attempt = 0
            while attempt < 40 && !settledStates.contains(state(of: journey)) {
                try await Task.sleep(nanoseconds: 250_000_000)
                journey = try await layerx.humanJourneyGet(pathParameters: ["journey_id": journeyID])
                attempt += 1
            }
            token.destroy()

            var report: [String: JSONValue] = [
                "journey_id": .string(journeyID),
                "state": .string(state(of: journey)),
                "receipts": .array(receipts(in: journey)),
            ]
            if let refusal = journey.objectValue?["refusal"]?.objectValue {
                report["refused_by"] = refusal["refused_by"] ?? .null
                report["money_left"] = refusal["money_left"] ?? .null
            }
            emit(report)
            exit(completedStates.contains(state(of: journey)) ? 0 : 2)
        } catch let error as PlatformSDKError {
            fail("refused: \(error.code.rawValue)")
        } catch {
            fail("refused: transport-failure")
        }
    }
}

File: platform/docs/samples/first-payment-jvm/src/main/java/com/sidiora/layerx/docs/FirstPayment.java
package com.sidiora.layerx.docs;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.sidiora.layerx.sdk.HttpProductionTransport;
import com.sidiora.layerx.sdk.IdempotencyKey;
import com.sidiora.layerx.sdk.PlatformSdkException;
import com.sidiora.layerx.sdk.ProductionClient;
import com.sidiora.layerx.sdk.SecretBytes;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class FirstPayment {
    private static final Set<String> SETTLED = Set.of("done", "done-finalised", "refused");
    private static final Set<String> COMPLETED = Set.of("done", "done-finalised");
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private FirstPayment() {}

    public static void main(String[] arguments) throws Exception {
        String apiUrl = required("LAYERX_API_URL");
        String apiToken = required("LAYERX_API_TOKEN");
        String source = required("LAYERX_SOURCE");
        String destination = required("LAYERX_DESTINATION");
        Map<String, String> money = Map.of("amount", required("LAYERX_AMOUNT"), "currency", required("LAYERX_CURRENCY"));
        String paymentKey = required("LAYERX_PAYMENT_KEY");

        // layerx:begin integration
        var credential = new HttpProductionTransport.BearerCredential(new SecretBytes(apiToken.getBytes(StandardCharsets.UTF_8)));
        var layerx = new ProductionClient(HttpProductionTransport.create(URI.create(apiUrl), URI.create(apiUrl), credential));
        var quote = layerx.human("move.quote", Map.of("source", source, "destination", destination, "money", money),
            JsonNode.class, ProductionClient.Options.none()).toCompletableFuture().join();
        var journey = layerx.human("move.commit", Map.of("quote_id", quote.path("quote_id").asText()), JsonNode.class,
            ProductionClient.Options.idempotent(new IdempotencyKey(paymentKey))).toCompletableFuture().join();
        // layerx:end integration

        JsonNode settled = awaitSettlement(layerx, journey);
        credential.close();
        System.out.println(report(settled));
        if (!COMPLETED.contains(settled.path("state").asText())) {
            System.exit(2);
        }
    }

    private static JsonNode awaitSettlement(ProductionClient layerx, JsonNode journey) throws InterruptedException {
        JsonNode current = journey;
        for (int attempt = 0; attempt < 40 && !SETTLED.contains(current.path("state").asText()); attempt += 1) {
            Thread.sleep(250L);
            current = layerx.human("journey.get", Map.of(), JsonNode.class,
                new ProductionClient.Options(null, Map.of("journey_id", current.path("journey_id").asText())))
                .toCompletableFuture().join();
        }
        return current;
    }

    private static String report(JsonNode journey) {
        ObjectNode report = MAPPER.createObjectNode();
        report.put("journey_id", journey.path("journey_id").asText());
        report.put("state", journey.path("state").asText());
        var receipts = report.putArray("receipts");
        for (JsonNode evidence : journey.path("evidence")) {
            if ("layerx-receipt".equals(evidence.path("class").asText())) {
                receipts.add(evidence.path("evidence_id").asText());
            }
        }
        if (journey.hasNonNull("refusal")) {
            report.put("refused_by", journey.path("refusal").path("refused_by").asText());
            report.put("money_left", journey.path("refusal").path("money_left").asBoolean());
        }
        return report.toString();
    }

    private static String required(String name) {
        String value = System.getenv(name);
        if (value == null || value.isEmpty()) {
            throw new IllegalStateException("first-payment-jvm: missing " + name);
        }
        return value;
    }
}
